File: android/lib/core/sync_indexer.dart
import 'dart:io';

import 'package:crypto/crypto.dart';

import 'sync_model.dart';

class SyncIndexer {
  static Future<SyncManifestMessage> generateManifest(
    String rootPath, [
    String folderId = 'default',
    bool computeHashOnDemand = false,
  ]) async {
    final manifest = SyncManifestMessage(
      folderId: folderId,
      timestamp: DateTime.now().millisecondsSinceEpoch,
      entries: [],
    );

    final dir = Directory(rootPath);
    if (!dir.existsSync()) {
      return manifest;
    }

    final normalizedRoot = rootPath
        .replaceAll('\\', '/')
        .replaceAll(RegExp(r'/+$'), '');
    final entities = dir.listSync(recursive: true, followLinks: false);

    for (final entity in entities) {
      if (entity is File) {
        final path = entity.path.replaceAll('\\', '/');
        final relative = path.startsWith('$normalizedRoot/')
            ? path.substring(normalizedRoot.length + 1)
            : path;

        final stat = entity.statSync();
        String hash = '';
        if (computeHashOnDemand) {
          final bytes = await entity.readAsBytes();
          hash = sha256.convert(bytes).toString().toLowerCase();
        }

        manifest.entries.add(
          SyncFileEntry(
            relativePath: relative,
            size: stat.size,
            modifiedTime: stat.modified.millisecondsSinceEpoch,
            sha256: hash,
          ),
        );
      }
    }

    return manifest;
  }

  static SyncDifference calculateDifference(
    SyncManifestMessage localManifest,
    SyncManifestMessage remoteManifest,
  ) {
    final needPull = <SyncFileEntry>[];
    final needPush = <SyncFileEntry>[];
    final upToDate = <SyncFileEntry>[];

    final localMap = {
      for (final e in localManifest.entries) e.relativePath.toLowerCase(): e,
    };
    final remoteMap = {
      for (final e in remoteManifest.entries) e.relativePath.toLowerCase(): e,
    };

    // Check remote entries against local
    for (final entry in remoteManifest.entries) {
      final key = entry.relativePath.toLowerCase();
      final local = localMap[key];

      if (local == null) {
        // Remote has it, local does not -> need pull
        needPull.add(entry);
      } else {
        // Tier 1: Fast metadata match (Size and MTime within 1s tolerance)
        final metadataMatch =
            local.size == entry.size &&
            (local.modifiedTime - entry.modifiedTime).abs() < 1000;

        // Tier 2: Hash match if both hashes are non-empty
        final hashMatch =
            local.sha256.isNotEmpty &&
            entry.sha256.isNotEmpty &&
            local.sha256.toLowerCase() == entry.sha256.toLowerCase() &&
            local.size == entry.size;

        if (metadataMatch || hashMatch) {
          upToDate.add(local);
        } else {
          // Last-Write-Wins conflict arbitration
          if (entry.modifiedTime > local.modifiedTime) {
            needPull.add(entry);
          } else if (local.modifiedTime > entry.modifiedTime) {
            needPush.add(local);
          } else {
            // Deterministic tie-break by SHA-256
            if (entry.sha256.compareTo(local.sha256) > 0) {
              needPull.add(entry);
            } else {
              needPush.add(local);
            }
          }
        }
      }
    }

    // Check local entries not in remote -> need push
    for (final local in localManifest.entries) {
      final key = local.relativePath.toLowerCase();
      if (!remoteMap.containsKey(key)) {
        needPush.add(local);
      }
    }

    return SyncDifference(
      needPull: needPull,
      needPush: needPush,
      upToDate: upToDate,
    );
  }
}

File: windows/Hinge.Core/SyncIndexer.cs
using System.Security.Cryptography;

namespace Hinge.Core;

public static class SyncIndexer
{
    public static async Task<SyncManifestMessage> GenerateManifestAsync(
        string rootPath,
        string folderId = "default",
        bool computeHashOnDemand = false)
    {
        var manifest = new SyncManifestMessage
        {
            FolderId = folderId,
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        };

        if (!Directory.Exists(rootPath))
        {
            return manifest;
        }

        var dirInfo = new DirectoryInfo(rootPath);
        var files = dirInfo.GetFiles("*", SearchOption.AllDirectories);

        foreach (var file in files)
        {
            // Normalize path to relative with forward slashes
            string relative = Path.GetRelativePath(rootPath, file.FullName).Replace('\\', '/');
            long size = file.Length;
            long modTime = new DateTimeOffset(file.LastWriteTimeUtc).ToUnixTimeMilliseconds();

            string sha256 = computeHashOnDemand ? await ComputeSha256Async(file.FullName) : string.Empty;

            manifest.Entries.Add(new SyncFileEntry
            {
                RelativePath = relative,
                Size = size,
                ModifiedTime = modTime,
                Sha256 = sha256,
                PlatformFileIdentity = null
            });
        }

        return manifest;
    }

    public static SyncDifference CalculateDifference(SyncManifestMessage localManifest, SyncManifestMessage remoteManifest)
    {
        var diff = new SyncDifference();

        var localMap = localManifest.Entries.ToDictionary(e => e.RelativePath, e => e, StringComparer.OrdinalIgnoreCase);
        var remoteMap = remoteManifest.Entries.ToDictionary(e => e.RelativePath, e => e, StringComparer.OrdinalIgnoreCase);

        // Check remote entries against local
        foreach (var (remotePath, remoteEntry) in remoteMap)
        {
            if (!localMap.TryGetValue(remotePath, out var localEntry))
            {
                // Remote has it, local does not -> need to pull from remote
                diff.NeedPull.Add(remoteEntry);
            }
            else
            {
                // Tier 1: Fast metadata match (Size and MTime within 1s tolerance for cross-FS clock precision)
                bool metadataMatch = localEntry.Size == remoteEntry.Size &&
                                     Math.Abs(localEntry.ModifiedTime - remoteEntry.ModifiedTime) < 1000;

                // Tier 2: Hash match if hashes were pre-calculated
                bool hashMatch = !string.IsNullOrEmpty(localEntry.Sha256) &&
                                 !string.IsNullOrEmpty(remoteEntry.Sha256) &&
                                 string.Equals(localEntry.Sha256, remoteEntry.Sha256, StringComparison.OrdinalIgnoreCase) &&
                                 localEntry.Size == remoteEntry.Size;

                if (metadataMatch || hashMatch)
                {
                    diff.UpToDate.Add(localEntry);
                }
                else
                {
                    // Conflict / modification: Last-Write-Wins
                    if (remoteEntry.ModifiedTime > localEntry.ModifiedTime)
                    {
                        diff.NeedPull.Add(remoteEntry);
                    }
                    else if (localEntry.ModifiedTime > remoteEntry.ModifiedTime)
                    {
                        diff.NeedPush.Add(localEntry);
                    }
                    else
                    {
                        // Deterministic tie-break by SHA-256 or size
                        if (string.CompareOrdinal(remoteEntry.Sha256, localEntry.Sha256) > 0)
                        {
                            diff.NeedPull.Add(remoteEntry);
                        }
                        else
                        {
                            diff.NeedPush.Add(localEntry);
                        }
                    }
                }
            }
        }

        // Check local entries not in remote -> need to push
        foreach (var (localPath, localEntry) in localMap)
        {
            if (!remoteMap.ContainsKey(localPath))
            {
                diff.NeedPush.Add(localEntry);
            }
        }

        return diff;
    }

    private static async Task<string> ComputeSha256Async(string filePath)
    {
        using var stream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read, 64 * 1024, useAsync: true);
        using var sha = SHA256.Create();
        byte[] hash = await sha.ComputeHashAsync(stream);
        return Convert.ToHexString(hash).ToLowerInvariant();
    }
}
